### PR TITLE
Fixing external communicator bugs

### PR DESCRIPTION
**Pull request by abhinavsns**

### CODE REVIEW

File: src/VCluster/VCluster_base.hpp [hunk 1]
```diff
@@ -398,18 +398,19 @@ class Vcluster_base
                 void *tag_ub_v;
                 int tag_ub;
 
-                MPI_Comm_get_attr(MPI_COMM_WORLD, MPI_TAG_UB, &tag_ub_v, &flag);
-                tag_ub = *(int*)tag_ub_v;
+               MPI_Comm_get_attr(ext_comm, MPI_TAG_UB, &tag_ub_v, &flag);
```
Comment (syaskovets) [inline src/VCluster/VCluster_base.hpp]: Please fix indentation

File: src/VCluster/VCluster.cpp [hunk 2]
```diff
@@ -64,9 +64,8 @@ void openfpm_init_vcl(int *argc, char ***argv, MPI_Comm ext_comm)
 {
 
 #if defined (ENABLE_NUMERICS) && defined (HAVE_PETSC)
-
+    PETSC_COMM_WORLD = ext_comm;
```
Comment (syaskovets) [inline src/VCluster/VCluster.cpp]: Please fix indentation

File: src/VCluster/VCluster_base.hpp [hunk 3]
```diff
@@ -398,18 +398,19 @@ class Vcluster_base
...
 
                 if (flag == true)
                 {
+                    tag_ub = *(int*)tag_ub_v;
```
Comment (syaskovets) [inline src/VCluster/VCluster_base.hpp]: Please fix indentation

File: src/VCluster/VCluster_base.hpp (hunk 1 above)
Comment (abhinavsns) [reply src/VCluster/VCluster_base.hpp]: done

File: src/VCluster/VCluster.cpp (hunk 2 above)
Comment (abhinavsns) [reply src/VCluster/VCluster.cpp]: done

File: src/VCluster/VCluster_base.hpp (hunk 3 above)
Comment (abhinavsns) [reply src/VCluster/VCluster_base.hpp]: done

**Comment by syaskovets** [comment]
This line https://github.com/mosaic-group/openfpm_vcluster/blob/develop/src/VCluster/VCluster.cpp#L67 fails on CUDA. When executed in debugger PETSC_COMM_WORLD is being assigned with ompi_mpi_comm_world but PETSC_COMM_WORLD is assigned with ompi_mpi_comm_null. The assignment later to PETSC_COMM_WORLD leads to erroneous behaviour in a multiple-process scenario. Reverting the commit 4d5d672a3cb8a34f6e40529294e510fb7b6c71f6. Setup to reproduce:
(Open MPI) 4.1.6
GCC 9.4.0
NVIDIA DRIVER version      : 560.28.03
nvcc V11.4.152

**Comment by abhinavsns** [comment]
Which test is causing the issue. Also I do not understand the steps to reproduce. Does it only happen in the debugger? 
Based on the documentation here, that line should be the right one: https://petsc.org/release/manualpages/Sys/PetscInitialize/
Are you sure that is the issue or it is caused by something else?

**Comment by syaskovets** [comment]
I was actually referring to this line `PETSC_COMM_WORLD = ext_comm;`. There is no problem with `PetscInitialize`. 
For example this test fails _dcpse_op_suite_tests_cu/dcpse_op_solver_. Yes, I am sure cause when commented the issue goes away. No, it happens when run normally, I used debugger to find out more about the issue

**Comment by abhinavsns** [comment]
Sorry I think you missed the petsc documentation:

> If you wish PETSc code to run ONLY on a subcommunicator of MPI_COMM_WORLD, create that communicator first and assign it to [PETSC_COMM_WORLD](https://petsc.org/release/manualpages/Sys/PETSC_COMM_WORLD/) BEFORE calling [PetscInitialize](https://petsc.org/release/manualpages/Sys/PetscInitialize/)(). Thus if you are running a four process job and two processes will run PETSc and have [PetscInitialize](https://petsc.org/release/manualpages/Sys/PetscInitialize/)() and PetscFinalize() and two process will not, then do this. If ALL processes in the job are using [PetscInitialize](https://petsc.org/release/manualpages/Sys/PetscInitialize/)() and [PetscFinalize](https://petsc.org/release/manualpages/Sys/PetscFinalize/)() then you don’t need to do this, even if different subcommunicators of the job are doing different things with PETSc.

Maybe it is a petsc bug from petsc CUDA solvers? In this case, I recommend putting it under `#ifdef NVCC`
Nevertheless I will try reproducing it and talk to petsc people. Thanks!

**Pull request merged**